drivers/can/can_rcar.c

```diff
@@ -0,0 +1,1044 @@
+		LOG_DBG("Bus-off entry interrupt\n");
+		sys_write8(RCAR_CAN_IER_ERSIE, config->reg_addr + RCAR_CAN_IER);
+		/* Clear interrupt condition */
+		sys_write8((uint8_t)~RCAR_CAN_EIFR_BOEIF,
+			   config->reg_addr + RCAR_CAN_EIFR);
+		can_rcar_state_change(dev, CAN_BUS_OFF);
+	}
+	if (eifr & RCAR_CAN_EIFR_ORIF) {
+		LOG_DBG("Receive overrun error interrupt\n");
+		sys_write8((uint8_t)~RCAR_CAN_EIFR_ORIF,
+			   config->reg_addr + RCAR_CAN_EIFR);
+	}
+	if (eifr & RCAR_CAN_EIFR_OLIF) {
+		LOG_DBG("Overload Frame Transmission error interrupt\n");
+		sys_write8((uint8_t)~RCAR_CAN_EIFR_OLIF,
+			   config->reg_addr + RCAR_CAN_EIFR);
+	}
+	if (eifr & RCAR_CAN_EIFR_BLIF) {
+		LOG_DBG("Bus lock detected interrupt\n");
+		sys_write8((uint8_t)~RCAR_CAN_EIFR_BLIF,
+			   config->reg_addr + RCAR_CAN_EIFR);
+	}
+}
+
+static void can_rcar_rx_filter(struct can_rcar_data *data,
+			       const struct zcan_frame *msg)
+{
+	struct zcan_frame tmp_msg;
+	uint8_t i;
+
+	k_mutex_lock(&data->rx_mutex, K_FOREVER);
+	for (i = 0; i < CONFIG_CAN_RCAR_MAX_FILTER; i++) {
+		if (data->rx_callback[i] == NULL) {
+			continue;
+		}
+
+		if (!can_utils_filter_match(msg,
+					    &data->filter[i])) {
+			continue; /* filter did not match */
+		}
+		/* Make a temporary copy in case the user
+		 * modifies the message.
+		 */
+		tmp_msg = *msg;
+		data->rx_callback[i](&tmp_msg, data->rx_callback_arg[i]);
+	}
+	k_mutex_unlock(&data->rx_mutex);
+}
+
+static void can_rcar_rx_handler(const struct device *dev)
+{
+	struct can_rcar_data *data = DEV_CAN_DATA(dev);
+	const struct can_rcar_cfg *config = DEV_CAN_CFG(dev);
+	struct zcan_frame msg;
+	uint32_t val;
+	int i;
+
+	val = sys_read32(config->reg_addr + RCAR_CAN_MB_60);
+	if (val & RCAR_CAN_MB_IDE) {
+		msg.id_type = CAN_EXTENDED_IDENTIFIER;
+		msg.id = val & RCAR_CAN_MB_EID_MASK;
+	} else {
+		msg.id_type = CAN_STANDARD_IDENTIFIER;
+		msg.id = (val & RCAR_CAN_MB_SID_MASK) >> RCAR_CAN_MB_SID_SHIFT;
+	}
+
+	if (val & RCAR_CAN_MB_RTR) {
+		msg.rtr = CAN_REMOTEREQUEST;
+	} else {
+		msg.rtr = CAN_DATAFRAME;
+	}
+
+	msg.dlc = sys_read16(config->reg_addr
+			     + RCAR_CAN_MB_60 + RCAR_CAN_MB_DLC_OFFSET) & 0xF;
+
+	/* Be paranoid doc states that any value greater than 8
+	 * should be considered as 8 bytes.
+	 */
+	if (msg.dlc > CAN_MAX_DLC) {
+		msg.dlc = CAN_MAX_DLC;
+	}
+
+	for (i = 0; i < msg.dlc; i++) {
+		msg.data[i] = sys_read8(config->reg_addr
+				+ RCAR_CAN_MB_60 + RCAR_CAN_MB_DATA_OFFSET + i);
+	}
+#if defined(CONFIG_CAN_RX_TIMESTAMP)
+	/* read upper byte */
+	msg.timestamp = sys_read8(config->reg_addr +
+				  RCAR_CAN_MB_60 + RCAR_CAN_MB_TSH_OFFSET);
+	msg.timestamp = msg.timestamp << 8;
+	/* and then read lower byte */
+	msg.timestamp |= sys_read8(config->reg_addr +
+				   RCAR_CAN_MB_60 + RCAR_CAN_MB_TSL_OFFSET);
+#endif
+	/* Increment CPU side pointer */
+	sys_write8(0xff, config->reg_addr + RCAR_CAN_RFPCR);
+
+	can_rcar_rx_filter(data, &msg);
+}
+
+static void can_rcar_int_thread(const struct device *dev)
```

Comment:
Is there a reason why you need a thread?
The API is defined that the callback is an ISR (Except for MCP2515).

Reply:
Idea there is to protect concurrent access of rx_callback[], and rx_callback_arg by locking data->rx_mutex.
Also for the Tx path I do lock data->inst_mutex to avoid isr, accessing and other threads accessing data->tx_head at the same time.

Reply:
You are writing from a thread and reading from an ISR.
The ISR cannot get interrupted.
If you have a member in the struct that can be written atomic and defines the validity, you don't need to lock.
Just write this marker at the end of an attach and at the beginning of a detach.

Reply:
For the ring buffer: writing the head from the ISR and tail from the thread can be done threads safe too.
You just need to add a `compiler_barier();` then moving hand/tail.

Reply:
Thanks for your suggestion, I did post a new version which get rid of this thread, and implements compiler_barrier as suggested.